File: src/bmorphism-mcp/typescript/test_server.ts
import { Server } from './server';

describe('Clojure Metaphors in TypeScript', () => {
  let server: Server;

  beforeEach(() => {
    server = new Server();
  });

  test('Immutable Foundations', () => {
    /* Like bedrock beneath a mountain, immutability provides stability */
    expect(server.getState()).toBe('initial');
    expect(() => {
      // @ts-expect-error - Testing immutability
      server.state = 'modified';
    }).toThrow();
  });

  test('Functional Streams', () => {
    /* Like a river's current, data flows without side effects */
    const result = server.processData([1, 2, 3]);
    expect(result).toEqual([2, 3, 4]);
    expect(server.getData()).toEqual([1, 2, 3]);
  });

  test('Composition as Symbiosis', () => {
    /* Like symbiotic relationships, functions work together harmoniously */
    const composed = server.composeFunctions(
      (x: number) => x + 1,
      (x: number) => x * 2
    );
    expect(composed(3)).toBe(8); // (3 + 1) * 2
  });

  test('Persistent Growth', () => {
    /* Like a tree adding rings, we grow without destroying the past */
    const original = server.getData();
    const modified = server.updateData(4);
    expect(modified).toEqual([1, 2, 3, 4]);
    expect(server.getData()).toEqual(original);
  });

  test('Higher-Order Wisdom', () => {
    /* Like a wise elder passing down knowledge, functions teach functions */
    const transform = server.createTransformer((x: number) => x * 2);
    expect(transform([1, 2, 3])).toEqual([2, 4, 6]);
  });
});

File: src/bmorphism-mcp/python/test_server.py
import unittest
from server import Server

class TestClojureMetaphors(unittest.TestCase):
    """A garden where Python blossoms with Clojure's wisdom"""
    
    def test_immutable_roots(self):
        """Like a tree's roots, immutability provides stability"""
        server = Server()
        # Test immutability patterns
        self.assertEqual(server.get_state(), "initial")
        with self.assertRaises(AttributeError):
            server.state = "modified"
            
    def test_functional_flow(self):
        """Rivers flow without changing their banks - pure functions"""
        server = Server()
        result = server.process_data([1, 2, 3])
        self.assertEqual(result, [2, 3, 4])
        # Original data remains unchanged
        self.assertEqual(server.get_data(), [1, 2, 3])
        
    def test_composition_as_ecosystem(self):
        """An ecosystem thrives through composition"""
        server = Server()
        composed = server.compose_functions(
            lambda x: x + 1,
            lambda x: x * 2
        )
        self.assertEqual(composed(3), 8)  # (3 + 1) * 2
        
    def test_persistent_data_structures(self):
        """Like rings in a tree, we add without destroying"""
        server = Server()
        original = server.get_data()
        modified = server.update_data(4)
        self.assertEqual(modified, [1, 2, 3, 4])
        self.assertEqual(server.get_data(), original)

if __name__ == '__main__':
    unittest.main()
